app/src/main/java/com/github/h3lp3rs/h3lp/messaging/Conversation.kt

```diff
@@ -16,26 +28,140 @@ class Conversation(
     private val currentMessenger: Messenger
 ) {
     private val database = databaseOf(MESSAGES)
+    private var publicKey: PublicKey? = null
+    private val allMessages: HashMap<String, String> = HashMap()
+    private val keyAlias = "CONVERSATION_KEY_$conversationId"
+    private val TRANSFORMATION = "RSA/ECB/OAEPwithSHA-1andMGF1Padding"
+    init {
+        //TODO
+        // Generate key pair
+        // share public key on DB
+
+
+        val keyStore = KeyStore.getInstance(ANDROID_KEY_STORE).apply { load(null) }
+        val alias = keyStore.aliases().toList()
+
+        if(!alias.contains(keyAlias)) {
+            val kpg =
+                KeyPairGenerator.getInstance(KeyProperties.KEY_ALGORITHM_RSA, ANDROID_KEY_STORE)
+
+            kpg.initialize(
+                KeyGenParameterSpec.Builder(
+                    keyAlias, KeyProperties.PURPOSE_ENCRYPT
+                            or KeyProperties.PURPOSE_DECRYPT
+                )
+                    .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_RSA_OAEP)
+                    .setDigests(KeyProperties.DIGEST_SHA1)
+                    .build()
+            )
+
+            val kp = kpg.generateKeyPair()
+
+            // send public key to the database
+            val encodedPublicKey = Base64.encodeToString(kp.public.encoded, Base64.DEFAULT)
+            database.setString(conversationId + "/KEYS/" + currentMessenger.name, encodedPublicKey)
+        }
+
+    }
 
     /**
      * Sends a message from the current user to the database
      * @param messageText The message text
      */
     fun sendMessage(messageText: String) {
-        val message = Message(currentMessenger, messageText)
-        database.addToObjectsListConcurrently(conversationId, Message::class.java, message)
+        // retrieve public key from Bob
+        if (publicKey == null) {
+            val path = conversationId + "/KEYS/" + Messenger.values()[(currentMessenger.ordinal + 1) % 2].name
+            database.getString(path).thenApply {
+                val decodedKey = Base64.decode(it, Base64.DEFAULT)
+                val keySpecs = X509EncodedKeySpec(decodedKey)
+                publicKey = KeyFactory.getInstance("RSA").generatePublic(keySpecs)
+                sendEncryptedMessage(publicKey!!, messageText) //TODO CHANGE THE !!
+            }
+        } else {
+            sendEncryptedMessage(publicKey!!, messageText)
+        }
     }
 
+    /**
+     *
+     */
+    private fun sendEncryptedMessage(publicKey: PublicKey, message: String) {
```

Comment:
This method name seems to hint to the fact that sendMessage doesn't actually send encrypted messages, maybe changing its name to something like encryptAndSendMessage would be clearer